File: backend/internal/domain/booking.go
package domain

import "time"

// Booking statuses. These are stored in the legacy `request.status` varchar
// (no DB constraint); `confirmed` is a mobile addition to the legacy set.
const (
	BookingPending             = "in_progress"
	BookingPendingVerification = "pending_verification"
	BookingConfirmed           = "confirmed"
	BookingActive              = "active"
	BookingCancelled           = "cancelled"
)

// Booking is a rental request. It maps onto the legacy `request` table.
// Nullable legacy columns are flattened to empty strings / nil pointers.
type Booking struct {
	ID              int32
	HouseID         int32
	UserID          int32
	GuestID         string
	Email           string
	Name            string
	Surname         string
	Lastname        string
	Count           int32
	Message         string
	Phone           string
	PhoneNormalized string
	StartDate       time.Time
	EndDate         *time.Time
	Status          string
	CreatedAt       time.Time
	UpdatedAt       time.Time
	ConfirmedAt     *time.Time
	RejectionReason string

	// House is a lightweight summary of the booked listing, populated by
	// list/detail queries (nil for the create/transition results).
	House *BookingHouse

	// Guest is the renter's profile data from the user table.
	// Populated only on the detail query (GET /requests/{id}).
	Guest *BookingGuest
}

// BookingHouse is the listing summary attached to a booking.
type BookingHouse struct {
	ID                int32
	OwnerID           int32
	Street            string
	HouseNumber       string
	// NumberRoom is the apartment/room number (кв. N). Hidden in DTOs until
	// the booking is confirmed or active to protect the guest's privacy pre-approval.
	NumberRoom        string
	City              string
	Price             int32
	CoverPath         string
	OwnerName         string
	OwnerSurname      string
	OwnerPatronymic   string
	OwnerPhone        string
	OwnerAvatarURL    string
	OwnerRating       float64
	OwnerReviewsCount int32
	OwnerIsVerified   bool
}

// BookingGuest is the profile of the user who created the booking,
// fetched from the user table on the detail endpoint.
type BookingGuest struct {
	Name         string
	Surname      string
	Patronymic   string
	AvatarURL    string
	Phone        string
	IsVerified   bool
	Rating       float64
	ReviewsCount int32
}

// NewBooking carries the validated fields needed to create a booking.
type NewBooking struct {
	HouseID         int32
	UserID          int32
	GuestID         string
	Email           string
	Name            string
	Surname         string
	Lastname        string
	Count           int32
	Message         string
	Phone           string
	PhoneNormalized string
	StartDate       time.Time
	EndDate         *time.Time
}

File: mobile/src/types/booking.ts
/**
 * Booking types mirror the backend DTOs (snake_case kept 1:1 with the wire
 * format). See backend/internal/delivery/http/booking_handler.go.
 */

/** Legacy `request.status` values used on mobile. */
export type BookingStatus = 'in_progress' | 'confirmed' | 'cancelled' | 'pending_verification';

/** Brief listing card embedded in booking list/detail responses. */
export interface BookingHouse {
  id: number;
  owner_id: number;
  address: string;
  city: string;
  /** Price per night, in rubles. */
  price: number;
  cover_url: string;
  owner_name?: string;
  owner_surname?: string;
  owner_patronymic?: string;
  owner_phone?: string;
  owner_avatar_url?: string;
  owner_rating?: number;
  owner_reviews_count?: number;
  owner_is_verified?: boolean;
}

/** Guest profile from the user table, available on the detail endpoint. */
export interface BookingGuest {
  name: string;
  surname: string;
  patronymic?: string;
  avatar_url: string;
  phone: string;
  is_verified: boolean;
  rating: number;
  reviews_count: number;
}

export interface Booking {
  id: number;
  house_id: number;
  user_id: number;
  name: string;
  surname: string;
  lastname: string;
  /** Number of guests. */
  count: number;
  message: string;
  phone: string;
  /** Check-in date, `YYYY-MM-DD`. */
  start_date: string;
  /** Check-out date, `YYYY-MM-DD`; null for open-ended legacy rows. */
  end_date: string | null;
  status: string;
  rejection_reason: string;
  /** RFC3339 timestamp set when an owner confirms; null otherwise. */
  confirmed_at: string | null;
  created_at: string;
  updated_at: string;
  house?: BookingHouse;
  /** Guest profile from user table; only present on GET /requests/{id}. */
  guest?: BookingGuest;
}

export interface BookingsPage {
  items: Booking[];
  total: number;
  limit: number;
  offset: number;
}

/** Body of POST /api/v1/listings/{id}/requests. */
export interface CreateBookingBody {
  count: number;
  name: string;
  surname?: string;
  lastname?: string;
  phone: string;
  message?: string;
  /** `YYYY-MM-DD`. */
  start_date: string;
  /** `YYYY-MM-DD`; omit for an open-ended request. */
  end_date?: string;
  /** Guest email for OTP verification (guest-mode only). */
  email?: string;
}

